---
title: "Analysis Bacteria"
author: "Oriol Castellano"
output: html_document
---

```{r}
# Load required libraries
library(dplyr)
library(ggplot2)
library(stringr)
library(dplyr)
# BiocManager::install("phyloseq")
library(phyloseq)
library(vegan) # Alpha and beta diversity
library(tidyr)
library(amap) # dendogram
library(ctc) # convert to newick
library(dendextend) # colour dend
library(purrr)
library(pheatmap)
library(rstatix)
library(ggpubr)


#Load data
data <- read.csv("/home/uriii/Desktop/BIOINFORMATICS/INT_FDG/ANALYSIS/DATA/Metagenomics/KRAKEN/BACTERIA/DATA_BACTERIA.csv") # metadata + kraken_data
data <- read.csv("DATA_BACTERIA.csv") # metadata + kraken_data

# MAYBE I HAVE TO SET "D55" == "D49"

metadata <- read.csv("/home/uriii/Desktop/BIOINFORMATICS/INT_FDG/ANALYSIS/DATA/Metagenomics/KRAKEN/BACTERIA/metadata_BACTERIA.csv")
metadata <- read.csv("metadata_BACTERIA.csv")

kraken_data <- read.csv("/home/uriii/Desktop/BIOINFORMATICS/INT_FDG/ANALYSIS/DATA/Metagenomics/KRAKEN/BACTERIA/report_data_BACTERIA.csv")
kraken_data <- read.csv("report_data_BACTERIA.csv")

data$Day[data$Day == "55"] <- "49"

data <- data[data$Taxon_name != "unclassified", ]
data <- data[data$Taxon_name != "root", ]
cellular_organisms <- data[data$Taxon_name == "cellularorganisms", ]

bacteria <- data[data$Taxon_name == "Bacteria", ] # D
bacteria <- unique(bacteria)
bacteria <- na.omit(bacteria)

data <- data[data$Taxon_name != "Bacteria", ]
data <- data[data$Taxon_name != "cellular organisms", ]
data <- unique(data)



# Actino <- data[data$Taxon_name == "Actinobacillus", ]
# Actino <- na.omit(Actino)
# Actino <- Actino[, c(1,2,3,4,10,13,15)]
# write.csv(Actino, "actino.csv")
```

```{r}
mbacteria <- mean(bacteria$Percent)
mbacteria
```


# Alpha diversity
BETWEEN SAMPLES
```{r}
# IN ALL THE DATA -- BETWEEN SAMPLES
data_alpha <- data[, c(1, 11, 15)]

# Aggregate the values (summing counts for duplicate combinations --> I lose some samples due to this)
aggregated_data <- aggregate(Count ~ Sample + Taxon_name, data = data_alpha, sum)

# Pivot the data to wide format
wide_data <- spread(aggregated_data, Taxon_name, Count)

# Remove row names and set 'Sample' as a separate column
# wide_data <- data.frame(Sample = rownames(wide_data), wide_data)

# Replace NAs with 0
wide_data[is.na(wide_data)] <- 0

# Extract 'Sample' as row names
rownames(wide_data) <- wide_data$Sample
wide_data <- wide_data[, -1]
# wide_data$Sample.1 <- NULL

# Calculate alpha diversity (Shannon index) on WIDE DATA
alpha_div_w <- diversity(wide_data, index = "shannon")
summary(alpha_div_w)
alpha_div_w <- as.data.frame(alpha_div_w)
alpha_div_w$Samples <- rownames(alpha_div_w)

# data_alphadiv_w <- cbind(alpha_div_w, wide_data) # combine Alpha diversity indices 
kruskal1 <- kruskal.test(alpha_div_w$alpha_div_w ~ alpha_div_w$Samples) # Kruskal-Wallis
print(kruskal1) # the p-value is 0.4726, which is greater than 0.05 == is no significant difference in Shannon diversity between SAMPLES, based on the Kruskal-Wallis Test
```

BETWEEN GROUPS
```{r}
process_group <- function(group_data) {
  # Aggregate the values
  aggregated_group <- group_data %>%
    group_by(Sample, Taxon_name) %>%
    summarise(Count = sum(Count)) %>%
    ungroup()

  # Pivot the data to wide format
  wide_group <- spread(aggregated_group, Taxon_name, Count, fill = 0)

  # Calculate alpha diversity (Shannon index)
  alpha_group <- diversity(wide_group[, -1], index = "shannon")

  # Combine alpha diversity indices with wide data
  data_alpha_group <- cbind(alpha_group, wide_group)

  return(data_alpha_group)
}

# Using purrr to process each group
alpha_div_results <- data %>%
  group_split(Group) %>%
  map(process_group)

# Combine the results into a single dataset
alpha_div_dataset <- bind_rows(alpha_div_results, .id = "Group")
alpha_div_dataset[is.na(alpha_div_dataset)] <- 0
# group 1 = NTS_IP, group 2 = NTS_NTP, group 3 = NTS_TP, group 4 = TS_IP, group 5 = TS_NTP # alpha_div_dataset$Group[alpha_div_dataset$Group == "1"] <- "NTS_IP"    # SO SLOW

alpha_div_dataset <- alpha_div_dataset %>%
  rename(alpha_div = alpha_group) %>%
  mutate(Group = case_when(
    Group == "1" ~ "NTS_IP",
    Group == "2" ~ "NTS_NTP",
    Group == "3" ~ "NTS_TP",
    Group == "4" ~ "TS_IP",
    Group == "5" ~ "TS_NTP",
    TRUE ~ as.character(Group) # Keep other values as is
  ))

# Define the custom order
custom_order <- c("NTS_NTP", "TS_NTP", "NTS_TP", "NTS_IP", "TS_IP")

# Convert the 'Group' variable to a factor with the custom order, just to change the order in the boxplot
alpha_div_dataset$Group <- factor(alpha_div_dataset$Group, levels = custom_order)

# BOXPLOT
f1 <- ggplot(alpha_div_dataset, aes(x = Group, y = alpha_div, fill = Group)) +
  geom_boxplot() +
  labs(title = "Alpha diversity (Shannon index) in FUNGI at different groups", x = "Groups", y = "Shannon Diversity") +
  scale_fill_manual(name = "Group", values = c("green", "orange", "red", "blue", "purple")) + # ADJUST MANUALLY
  theme_classic()
ggsave("/home/uriii/Desktop/BIOINFORMATICS/INT_FDG/REPORT AND FDP/Figures/SuppMat/Bacteria/Alpha_BATERIA.svg", f1, height = 6, width = 10)

kruskal2 <- kruskal.test(alpha_div_dataset$alpha_div ~ alpha_div_dataset$Group)
print(kruskal2) # the p-value is 0.01798, which is lower than 0.05 == is no significant difference in Shannon diversity between groups based on the Kruskal-Wallis test. 

# Perform pairwise Wilcoxon tests with adjustment for multiple comparisons
pairwise_results <- alpha_div_dataset %>%
  pairwise_wilcox_test(alpha_div ~ Group, p.adjust.method = "bonferroni")

# Print the pairwise test results
print(pairwise_results)
```
# Beta diversity
Between samples
```{r}
# Beta diversity analysis
dist_matrix <- vegdist(wide_data, method = "bray")
print(dist_matrix)
write.table(dist_matrix, file = "bray_curtis_matrix.txt", sep = "\t", quote = FALSE)

# PCoA -- Multidimensional scaling (MDS) is a means of visualizing the level of similarity of individual cases of a dataset.
pcoa_result <- cmdscale(dist_matrix)  #  is similar in spirit to PCA but it takes a dissimilarity as input! A dissimilarity matrix shows the distance between every possible pair of objects.

# color by sample
pcoa_result <- as.data.frame(pcoa_result)
pcoa_result$Sample <- rownames(pcoa_result)
data_pcoa <- metadata[, c(1,2,6)]
pcoa <- merge(pcoa_result, data_pcoa, by = c("Sample"), all.x = TRUE)
pcoa$Day <- factor(pcoa$Day) ###########################################################
pcoa$Day[pcoa$Day == "55"] <- "49"

ggplot(data = pcoa, aes(x = pcoa[, 2], y = pcoa[, 3])) +
  geom_point(aes(col = Day, shape = Group)) + 
  labs(title = "PCoA Plot - Days", x = "PC1", y = "PC2", color = "Extraction Day") +
  scale_color_manual(name = "Day", values = c("blue", "red"),  labels = c("49", "21")) + # values = color_vector,
  theme_classic() 
  # stat_ellipse(aes(col = Day))
```
```{r}
pcoa_day21 <- pcoa %>%
  filter(Day %in% "21")
b_d21 <- ggplot(data = pcoa_day21, aes(x = pcoa_day21[, 2], y = pcoa_day21[, 3], col = Group)) +
  geom_point(aes(col = Group)) + 
  labs(x = "PC1", y = "PC2", color = "Group") +
  scale_color_manual(name = "Group", values = c("blue", "green", "red", "purple", "orange")) + # values = color_vector,
  scale_fill_manual(name = "Group", values = c("blue", "green", "red", "purple", "orange")) + # values = color_vector,
  theme_classic() + 
  stat_ellipse(geom = "polygon", aes(fill = Group), alpha = 0.1) # Instead of 
b_d21
ggsave("/home/uriii/Desktop/BIOINFORMATICS/INT_FDG/REPORT AND FDP/Figures/SuppMat/Bacteria/Bacteria_PCoA_D21.png", b_d21, height = 6, width = 10)



pcoa_day49 <- pcoa %>%
  filter(Day %in% "49")
b_d49 <- ggplot(data = pcoa_day49, aes(x = pcoa_day49[, 2], y = pcoa_day49[, 3], col = Group)) +
  geom_point(aes(col = Group)) + 
  labs(x = "PC1", y = "PC2", color = "Group") +
  scale_color_manual(name = "Group", values = c("blue", "green", "red", "purple", "orange")) + # values = color_vector,
  scale_fill_manual(name = "Group", values = c("blue", "green", "red", "purple", "orange")) + # values = color_vector,
  theme_classic() + 
  stat_ellipse(geom = "polygon", aes(fill = Group), alpha = 0.1) # if I remove the fill from aes(), disappear from the legend --> I want this bit filling by colour to
b_d49
ggsave("/home/uriii/Desktop/BIOINFORMATICS/INT_FDG/REPORT AND FDP/Figures/SuppMat/Bacteria/Bacteria_PCoA_D49.png", b_d49, height = 6, width = 10)

```

```{r}
# Divide the previous PCoA by group
# Plot PCoA for Day 21, faceted by Group
ggplot(data = pcoa_day21, aes(x = pcoa_day21[, 2], y = pcoa_day21[, 3], col = Group)) +
  geom_point() +
  labs(title = "PCoA Plot - Day 21", x = "PC1", y = "PC2", color = "Group") +
  scale_color_manual(name = "Group", values = c("blue", "green", "red", "purple", "orange")) +
  facet_wrap(~Group) +
  theme_classic() +
  stat_ellipse(geom = "polygon", aes(fill = Group), alpha = 0.1)


# Plot PCoA for Day 49, faceted by Group
ggplot(data = pcoa_day49, aes(x = pcoa_day49[, 2], y = pcoa_day49[, 3], col = Group)) +
  geom_point() +
  labs(title = "PCoA Plot - Day 49", x = "PC1", y = "PC2", color = "Group") +
  scale_color_manual(name = "Group", values = c("blue", "green", "red", "purple", "orange")) +
  facet_wrap(~Group) +
  theme_classic() +
  stat_ellipse(geom = "polygon", aes(fill = Group), alpha = 0.1)

```

# PERMANOVA
```{r}
# data_perm <- data[, c(1, 4)] # Group and sample
# # for wide_data
# perm_wide_data <- wide_data
# perm_wide_data$Sample <- rownames(wide_data)
# perm_wide_data <- merge(perm_wide_data, data_perm, by = c("Sample")) # merge to have also the group 
# # perm_wide_data <- cbind(perm_wide_data, data_perm, by = c("Sample"), all.x = T) # merge to have also the group
# perm_wide_data <- unique(perm_wide_data)
# 
# # Option 1
# rownames(perm_wide_data) <- perm_wide_data$Sample # samples
# perm_wide_data <- perm_wide_data[, -1]
# perm_wide_data$Group <- as.numeric(factor(perm_wide_data$Group))
# 
# permanova_All <- adonis2(perm_wide_data ~ Group, data = perm_wide_data, method = "bray")
# permanova_All # NO statistically significant similarities between groups or STATISTICALLY SIGNIFICANT DIFFERENCES BETWEEN GROUPS
# 
# data_perm <- data[, c(1, 6)] # Sample and Day
# # for wide_data
# perm_wide_data <- wide_data
# perm_wide_data$Sample <- rownames(wide_data)
# perm_wide_data <- merge(perm_wide_data, data_perm, by = c("Sample")) # merge to have also the group 
# # perm_wide_data <- cbind(perm_wide_data, data_perm, by = c("Sample"), all.x = T) # merge to have also the group
# perm_wide_data <- unique(perm_wide_data)
# 
# # Option 1
# rownames(perm_wide_data) <- perm_wide_data$Sample # samples
# perm_wide_data <- perm_wide_data[, -1]
# perm_wide_data$Day <- as.numeric(factor(perm_wide_data$Day))
# 
# permanova_All <- adonis2(perm_wide_data ~ Day, data = perm_wide_data, method = "bray")
# permanova_All # NO statistically significant similarities between groups OR STATISTICALLY SIGNIFICANT DIFFERENCES BETWEEN DAYS
# 
# # Option 2
# perm_result <- adonis2(dist_matrix ~ Group, data = metadata, permutations = 999)
# perm_result

# OPTION 1
# data_perm <- data[, c(1, 4, 6)] # Group and sample
# perm_wide_data <- wide_data
# perm_wide_data$Sample <- rownames(wide_data)
# perm_wide_data <- merge(perm_wide_data, data_perm, by = c("Sample")) # merge to have also the group
# perm_wide_data <- unique(perm_wide_data)
# perm_wide_data_21 <- perm_wide_data[perm_wide_data$Day == "21", ]
# rownames(perm_wide_data_21) <- perm_wide_data_21$Sample # samples
# perm_wide_data_21 <- perm_wide_data_21[, -1]
# perm_wide_data_21 <- perm_wide_data_21[, -which(names(perm_wide_data_21) == "Day")]
# perm_wide_data_21$Group <- as.numeric(factor(perm_wide_data_21$Group))
# 
# permanova_21 <- adonis2(perm_wide_data_21 ~ Group, data = perm_wide_data_21, method = "bray")
# permanova_21 # LOWER THAN 0.05 -_> statistically different beta diversity at DAY 21
# 
# # OPTION 2
# abundance_data_21 <- perm_wide_data_21[, !(names(perm_wide_data_21) %in% c("Sample", "Group", "PigletID", "SowID", "Sampling_date", "Day", "Sow_treatment", "Piglet_treatment", "Antibiotic"))]
# group_labels_21 <- perm_wide_data_21$Group
# distance_matrix_21 <- vegdist(abundance_data_21, method = "bray")
# permanova_results_21 <- adonis(distance_matrix_21 ~ group_labels_21, data = perm_wide_data_21, permutations = 999)
# print(permanova_results_21)
# ```

# ```{r}
# # DAY 21
# data_perm <- data[, c(1, 4, 6)] # Group and sample
# perm_wide_data <- wide_data
# perm_wide_data$Sample <- rownames(wide_data)
# perm_wide_data <- merge(perm_wide_data, data_perm, by = c("Sample")) # merge to have also the group
# perm_wide_data <- unique(perm_wide_data)
# perm_wide_data_21 <- perm_wide_data[perm_wide_data$Day == "21", ]
# rownames(perm_wide_data_21) <- perm_wide_data_21$Sample # samples
# perm_wide_data_21 <- perm_wide_data_21[, -1]
# perm_wide_data_21 <- perm_wide_data_21[, -which(names(perm_wide_data_21) == "Day")]
# perm_wide_data_21$Group <- as.numeric(factor(perm_wide_data_21$Group))
# 
# permanova_21 <- adonis2(perm_wide_data_21 ~ Group, data = perm_wide_data_21, method = "bray")
# permanova_21 # ********* pvalue = 0.106 -------------- p HIGHER than 0.05 -_> similarities between groups at day 21 / NO DIFFERENCES
# ```
# 
# ```{r}
# # DAY 49
# data_perm <- data[, c(1, 4, 6)] # Group and sample
# perm_wide_data <- wide_data
# perm_wide_data$Sample <- rownames(wide_data)
# perm_wide_data <- merge(perm_wide_data, data_perm, by = c("Sample")) # merge to have also the group 
# perm_wide_data <- unique(perm_wide_data)
# perm_wide_data_49 <- perm_wide_data[perm_wide_data$Day == "49", ]
# rownames(perm_wide_data_49) <- perm_wide_data_49$Sample # samples
# perm_wide_data_49 <- perm_wide_data_49[, -1]
# perm_wide_data_49 <- perm_wide_data_49[, -which(names(perm_wide_data_49) == "Day")]
# perm_wide_data_49$Group <- as.numeric(factor(perm_wide_data_49$Group))
# 
# permanova_49 <- adonis2(perm_wide_data_49 ~ Group, data = perm_wide_data_49, method = "bray")
# permanova_49 # pvalue = 0.16 ,,,, p > 0.05 --> NO statistically significant DIFFERENCES between groups at day 49
```


# Heatmaps
```{r}
data_days <- data[, c(6, 10, 13, 15)] # NOW PERCENT. CHANGE TO COLUMN 11 TO USE COUNTS
rank_to_filter <- "G"
data_days <- data_days[data_days$Rank == rank_to_filter, ]

# Sum all Viruses that belong to the same day
# Aggregate the values
aggregated_data_days <- aggregate(Percent ~ Day + Taxon_name, data = data_days, sum)

# Pivot the data to wide format
wide_data_days <- spread(aggregated_data_days, Taxon_name, Percent)

# Replace NAs with 0
wide_data_days[is.na(wide_data_days)] <- 0

# Extract 'Day' as row names
rownames(wide_data_days) <- wide_data_days$Day
wide_data_days <- wide_data_days[, -1]
# wide_data_days$Day.1 <- NULL

# Now, to make a heatmap we ought to first transform the raw counts of reads to proportions within a sample
# wide_data_days <- wide_data_days/rowSums(wide_data_days) # NOT NECESSARY

# colorRampPalette is in the RColorBrewer package. This creates a colour palette that shades from light yellow to red in RGB space with 100 unique colours
scaleyellowred <- colorRampPalette(c("lightyellow", "red"), space = "rgb")(100)
# heatmap(as.matrix(prop_wide_data_days), Rowv = NA, Colv = NA, col = scaleyellowred)

# determine the maximum relative abundance for each column
maxab <- apply(wide_data_days, 2, max)

# remove the genera with less than 0.0001% as their maximum relative abundance
threshold <- 5
wide_data_days_filtered <- wide_data_days[, maxab >= threshold]

newnames <- lapply(
  colnames(wide_data_days_filtered),
  function(x) bquote(italic(.(x))))

# pheatmap(as.matrix(wide_data_days_filtered), col = scaleyellowred, margins = c(10, 3), labels_col = as.expression(newnames), main = "Heatmap - Day", angle_col = 45)

# Plot the heatmap using the filtered data
pheatmap(as.matrix(wide_data_days_filtered), col = scaleyellowred, margins = c(10, 3), 
         labels_col = as.expression(newnames), main = paste("Heatmap - Rank:", rank_to_filter), 
         angle_col = 45)
```
```{r}
plot_taxa_distribution <- function(data, ranks, variable) {
  # Filter rows based on specific taxonomic ranks
  data_filtered <- data %>%
    filter(Rank %in% ranks)

  # Create a dynamic label based on the variable
  y_label <- ifelse(variable == "Count", "Percent")

  # Barplot of Viral Taxa Distribution per Group
  ggplot(data_filtered, aes(x = Group, y = !!sym(variable), fill = Taxon_name)) +
    geom_bar(position = "stack", stat = "identity") +
    labs(x = 'Group', y = "Abundance", title = paste('Bacterial Taxa Distribution (' ,ranks, ') per Group')) +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
}
taxonomic_ranks <- c("C")# c("O", "F")
plot_variable <- "Count"  # Change to "Percent" if needed
taxa_per_group <- plot_taxa_distribution(data, taxonomic_ranks, plot_variable)
taxa_per_group
ggsave("/home/uriii/Desktop/BIOINFORMATICS/INT_FDG/REPORT AND FDP/Figures/SuppMat/Bacteria/Bacteria_Taxa_per_Group.png", taxa_per_group, height = 6, width = 10)
```
