---
title: "Data Preprocessing"
author: "Oriol Castellano"
output: html_document
---


```{r}
# Load required libraries
library(dplyr)
library(ggplot2)
library(stringr)
library(dplyr)
```


```{r}
# Directory containing Kraken reports
kraken_reports_dir <- "/home/uriii/Desktop/BIOINFORMATICS/INT_FDG/ANALYSIS/DATA/Metagenomics/KRAKEN/FUNGI/Reports"

# Function to read and process a Kraken report
process_kraken_report <- function(report_file) {
  # Read Kraken report into a data frame
  df <- read.table(report_file, header = FALSE, sep = "\t", quote = "", comment.char = "", stringsAsFactors = FALSE)
  
  # Add sample name as a column
  df$sample <- tools::file_path_sans_ext(basename(report_file))
  return(df)
}

# List to store data frames for each sample
list_reports <- list()


# Loop through each Kraken report
report_files <- list.files(kraken_reports_dir, pattern = "\\.report_F$", full.names = TRUE)
for (report_file in report_files) {
  # Process the Kraken report and add it to the list
  list_reports[[report_file]] <- process_kraken_report(report_file)
}

# Combine data frames into a single data frame
reports_df <- bind_rows(list_reports)  #rbind() or cbind()
colnames(reports_df) <- c("Percent", "Count", "Count2", "Rank", "Taxon_id", "Taxon_name", "Report")

# Replace spaces and tabs in the taxon_name column with an empty string
reports_df$Taxon_name <- str_trim(reports_df$Taxon_name, side = "left")
# reports_df$Taxon_name <- str_replace_all(reports_df$Taxon_name, "[[:space:]]", "") # with gsub --> reports_df$taxon_name <- gsub("[[:space:]]", "", reports_df$taxon_name)

# In order to remove the "report_V" string
# reports_df$Sample <- str_extract(reports_df$Report, "F\\d+") 
# reports_df$Sample <- reports_df$Sample + str_extract(reports_df$Report, "D\\d++")  --> MAKE A VARIABLE WITH F_____D_____

reports_df$PigletID <- str_extract(reports_df$Report, "F\\d+")
reports_df$Sampling_date <- str_extract(reports_df$Report, "D\\d+")
reports_df$Sample <- paste(reports_df$PigletID, reports_df$Sampling_date, sep = "_")



# Remove rows (= VIRUSES) where Percent is equal to 0.00
reports_df <- reports_df[reports_df$Percent != 0.00, ]
```
Remove F75_D21, as there's no more available data from days 49, 55, ..  --> not necessary yet


## Add metadata
NTS_NTP = Control 
```{r}
# ADD METADATA
metadata.a <- read.table("/home/uriii/Desktop/BIOINFORMATICS/INT_FDG/ANALYSIS/DATA/Metagenomics/KRAKEN/metadata/baro.wmgx.metadata.a.txt", header = TRUE, sep = "\t", quote = "", comment.char = "", stringsAsFactors = FALSE)
metadata.b <- read.table("/home/uriii/Desktop/BIOINFORMATICS/INT_FDG/ANALYSIS/DATA/Metagenomics/KRAKEN/metadata/baro.wmgx.metadata.b.txt", header = TRUE, sep = "\t", quote = "", comment.char = "", stringsAsFactors = FALSE)

# Combine metadata.a and metadata.b using bind_rows from dplyr
combined_metadata <- bind_rows(metadata.a, metadata.b)

# Remove the Piglet_inoculation column, and instead of, use the Piglet_treatment variable to determine if was "Inoculated"/"Ab"/"no"
combined_metadata$Piglet_inoculation[is.na(combined_metadata$Piglet_inoculation)] <- "No"
combined_metadata$Piglet_treatment[combined_metadata$Piglet_inoculation == "Inoculated"] <- "Inoculated"
combined_metadata$Piglet_treatment[is.na(combined_metadata$Piglet_treatment)] <- "No"

combined_metadata <- combined_metadata[, -3] # Remove Group_and_Day
combined_metadata <- combined_metadata[, -10] # Remove Piglet_inoculation

combined_metadata$Antibiotic[combined_metadata$Group == "TS_IP"] <- "yes"

combined_metadata$Sow_treatment[combined_metadata$Sow_treatment %in% c("T", "Ab")] <- "Yes"
combined_metadata$Sow_treatment[combined_metadata$Sow_treatment == "NT"] <- "No"

combined_metadata$Antibiotic[combined_metadata$Group == "TS_IP"] <- "yes"
combined_metadata$Antibiotic[is.na(combined_metadata$Antibiotic)] <- "no"
combined_metadata$Antibiotic[combined_metadata$Antibiotic == "yes"] <- "Yes"
combined_metadata$Antibiotic[combined_metadata$Antibiotic == "no"] <- "No"

combined_metadata$Sampling_date[combined_metadata$Day == "49"] <- "D49"
combined_metadata$Sampling_date[combined_metadata$Day == "55"] <- "D55"
```


# BIND METADATA AND REPORT DATA
```{r}
merged_data <- merge(combined_metadata, reports_df, by = c("Sample", "PigletID", "Sampling_date"), all.x = TRUE)
```
# SAVE DATAFRAMES IN LOCAL
```{r}
write.csv(combined_metadata, "/home/uriii/Desktop/BIOINFORMATICS/INT_FDG/ANALYSIS/DATA/Metagenomics/KRAKEN/FUNGI/metadata_FUNGI.csv", row.names = FALSE)
write.csv(reports_df, "/home/uriii/Desktop/BIOINFORMATICS/INT_FDG/ANALYSIS/DATA/Metagenomics/KRAKEN/FUNGI/report_data_FUNGI.csv", row.names = FALSE)
write.csv(merged_data, "/home/uriii/Desktop/BIOINFORMATICS/INT_FDG/ANALYSIS/DATA/Metagenomics/KRAKEN/FUNGI/DATA_FUNGI.csv", row.names = FALSE)
```

